### DISCUSSION TITLE
CFU vector operations

### DISCUSSION THREAD

**Discussion by jimmy-979** [stnolting/neorv32#446] [General]
Hello,
I would like to create some custom instructions to do 128-bit vector operations. The way I implement this functionality for now is to do something like this:

```c
uint32_t arr[4] = {0x0, 0x1, 0x2, x03};

neorv32_cfu_cmd0(0b0000000, arr[0], arr[1]); // load arr[0], arr[1] to vector_reg(0) LSB
neorv32_cfu_cmd0(0b0000001, arr[2], arr[3]); // load arr[2], arr[3] to vector_reg(0) MSB

// vector operations
// neorv32_cfu_cmd0(0b0000010, 0x0, 0x0); // do some rotation on 128-bit vector_reg(0)
```

Is it possible for CFU to access multiple memory elements based on some offset? 
For example the command
```c
neorv32_cfu_cmd0(0b0000000, arr[0], 4); 
```
could make the CFU access `arr[0], arr[0]+1, arr[0]+2, arr[0]+3` from memory immediately and concatenate them to my vector register file.

Thanks in advance,
Jimmy

**Comment by stnolting** [comment]
Hey there!

> Is it possible for CFU to access multiple memory elements based on some offset?

The CFU has no access to the CPU's memory port so it cannot read/write data from/to memory by itself. The CFU operates on the CPU's register file only. Hence, all source operands (= the elements of your array/vector) have to be _in_ the core's register file.

If your custom hardware module (= the CFU in this case) needs direct access to memory, you might be better off using a custom co-processor outside of the CPU/processor.

Another option to accelerate CFU data access would be to add additional read ports to the core's register file. For example, having one more read port you could use instructions with 3 source registers:

```c
neorv32_cfu_cmd0(0b0000000, arr[0], arr[1], arr[2]); // load arr[0] ... arr[2]
neorv32_cfu_cmd0(0b0000001, arr[3], arr[4], arr[5]); // load arr[3] ... arr[5]
```

Having more than 3 register file read ports would require a customization of the compiler because there are no standard RISC-V instructions with more than 3 register sources (at least not in the scalar extensions).

**Reply by stnolting** [reply]
I thought about this again. Maybe it would be a good idea to add a third source register option to the CFU so we can implement custom R4-type instructions (3 source registers, 1 destination register). This would make implementing common instruction like multiply-add (rs1*rs2+rs3) much more efficient. I am doing some experiments... maybe I'll do a PR. :wink:

Processing large arrays/vectors like in your application might benefit from that: ~50% faster since a single CFU instruction could then process 3 elements instead of just 2. However, the RISC-V `V` (vertical vectoring) or `P` (packed SIMD / horizontal vectoring).

Btw, I just found a great thesis where the NEORV32 processor gets a `V` vector extension upgrade: https://repositorio.comillas.edu/xmlui/handle/11531/64349 (Spanish). :rocket: :+1:

**Comment by stnolting** [comment]
In case you're interested - I started upgrading the CFU to support R4-type instructions (three source operands) in https://github.com/stnolting/neorv32/pull/449. The modifications are fully backwards-compatible.

**Comment by stnolting** [comment]
The CFU now supports R4-type instructions.

There are two custom opcodes left:

![Unbenannt](https://user-images.githubusercontent.com/22944758/205980930-88e5d1be-44b1-43c3-837d-3846361e98e2.png)

With these opcodes it would be possible to implement R5-type instructions with 4 source registers. As far as I know, there is no pre-defined RISC-V instruction word layout for this, but that is not really a problem. However, having 4 source registers and 1 destination register (each requiring 5-bits in the instruction) plus the 7-bit opcode would not leave any further bits to specify the actual operation.

**Reply by jimmy-979** [reply]
Hi,  

I’m trying to accelerate some cryptographic functions using the CFU and the custom instructions, so I thought that having 4 input registers could enable me, for example, to process the whole AES state in one command which would make it run faster. 

Probably this is a very specific case and if someone needed this functionality would use something like a custom coprocessor or try to add things to RISC-V vector extension.

**Reply by stnolting** [reply]
> I’m trying to accelerate some cryptographic functions using the CFU and the custom instructions, so I thought that having 4 input registers could enable me, for example, to process the whole AES state in one command which would make it run faster.

I think this a really good example for using custom instruction! :+1:

> Probably this is a very specific case and if someone needed this functionality would use something like a custom coprocessor or try to add things to RISC-V vector extension.

Other implementation options might be using a dedicated co-processor or implying the RISC-V cryptography ISA extension (`K`).

The `custom-2` and `custom-3` OPCODE are defined by RISC-V and are not used by the NEORV32 at all. So I think it would make sense to add them to the CFU providing the user maximum flexibility. But I am not sure about the r5-type format as there is no pre-defined layout for that.

![cfu_r4type_instruction](https://user-images.githubusercontent.com/22944758/206282014-279065e1-2901-45d2-b7e2-02ed01db2f03.png)

Basically, we could start with the r4-format and repurpose the `funct3` bits and the yet-unused bits `26:25` to encode the fourth source register `rs4` :thinking:

But then we have no bits left to specify the actual operation. By implementing two r5-type instructions using the `custom-2` and `custom-3` OPCODEs we would end up having just two r5-type instruction/operations. Of course an implementer could use other instruction types to "pre-configure" the actual operation before issuing those r5-type instructions...

Any thoughts?? :wink:

**Reply by stnolting** [reply]
R5-type instructions will be added with https://github.com/stnolting/neorv32/pull/452 😉
